crates/stages/src/stages/tx_lookup.rs

```diff
@@ -66,14 +66,35 @@ impl<DB: Database> Stage<DB> for TransactionLookupStage {
                 .unwrap_or_default()
         });
 
-        // Collect tranasctions for each body and insert the reverse lookup for hash -> tx_id.
+        // Collect transactions for each body
+        let mut tx_list = vec![];
         for body_entry in bodies {
             let (_, body) = body_entry?;
-            let transactions = tx_cursor.walk(body.start_tx_id)?.take(body.tx_count as usize);
+            let transactions = cursor_tx.walk(body.start_tx_id)?.take(body.tx_count as usize);
 
             for tx_entry in transactions {
                 let (id, transaction) = tx_entry?;
-                tx.put::<tables::TxHashNumber>(transaction.hash(), id)?;
+                tx_list.push((transaction.hash(), id));
+            }
+        }
+
+        let mut cursor_txhash = tx.cursor_write::<tables::TxHashNumber>()?;
+
+        // If the last element is smaller than our first, then we can just append into the DB.
+        // This probably only ever happens during sync, on the first table insertion.
+        let last = cursor_txhash.last()?;
+        let mut append = last.is_none();
+        if let (false, Some((last_txhash, _))) = (tx_list.is_empty(), last) {
+            append = last_txhash < tx_list[0].0;
+        }
+
+        // Sort before inserting the reverse lookup for hash -> tx_id.
+        tx_list.sort_by(|txa, txb| txa.0.cmp(&txb.0));
+        for (tx_hash, id) in tx_list {
+            if append {
+                cursor_txhash.append(tx_hash, id)?;
+            } else {
+                cursor_txhash.insert(tx_hash, id)?;
```

Comment:
Is this okay, `insert` inserts at current position of cursor, should we use `put` here?

Reply:
I think it's ok, because:
* `DBCursorRW` already uses `put()` in write operation. e.g, `upsert()`, `insert()` and `append()`. These methods' body are same except `WriteFlags`.
* The cursor position, before writing, is not meaningful except when flag is`WriteFlag::CURRENT`. 
* If writing operation is success then cursor position is change to point the new data.

ref: https://erthink.github.io/libmdbx/group__c__crud.html#ga5b8137591c45143c6d9439799be77136

code: 
https://github.com/paradigmxyz/reth/blob/c63f5ff9420acaf244473f1f53e7b38645d5503c/crates/storage/db/src/implementation/mdbx/cursor.rs#L177-L199

Reply:
I think we need to investigate `append()` because I think `DbCursorRW.append()` could behave unlike it's descrption.

> Appends the data to the end of the table. Consequently, the append operation will fail if the inserted key is less than the last table key

I guess, `WriteFlags::APPEND` is not related `key` but it is related with `value` like `UPSERT` and `NO_OVERWRITE`. I'll investigate it w/ mdbx source and share my learning.

https://github.com/paradigmxyz/reth/blob/c63f5ff9420acaf244473f1f53e7b38645d5503c/crates/storage/db/src/implementation/mdbx/cursor.rs#L193-L199

Reply:
I would recommend writing the test. Something like this: https://github.com/paradigmxyz/reth/blob/a55e7fd9b5b40acce3030a9366285947455138be/crates/storage/db/src/implementation/mdbx/mod.rs#L338
Setup would be values inserted `[ 1,3,5,7,9]` .
Test would be 
```
let mut cursor_txhash = tx.cursor_write::<tables::TxHashNumber>()?;
let last = cursor_txhash.last()?;
cursor_txhash.insert(2, 0)?;
cursor_txhash.insert(3, 0)?;
```

Reply:
fixed https://github.com/paradigmxyz/reth/pull/1161